pkg/maps/lbmap/maglev.go

```diff
@@ -37,126 +33,147 @@ const (
-			if err != nil {
-				return false, err
-			}
-			previousM := int(info.ValueSize) / int(unsafe.Sizeof(uint16(0)))
-			if option.Config.MaglevTableSize != previousM {
-				deleteMap = true
-			}
-		} else if err == io.EOF {
-			// The map is empty. To be on the safe side, remove it to avoid M
-			// mismatch. The removal is harmless, as no new entry can be
-			// created while the initialization of the map has not returned.
-			deleteMap = true
-		} else {
-			return false, err
+// MaybeInitMaglevMaps tries to open all already existing maglev BPF maps by
+// probing their table size from the inner map value size.
+func MaybeInitMaglevMaps() (uint32, error) {
+	var (
+		detectedTableSize uint32
+		err               error
+	)
+
+	map4Found, maglev4TableSize := MaglevOuterMapTableSize(MaglevOuter4MapName)
+	map6Found, maglev6TableSize := MaglevOuterMapTableSize(MaglevOuter6MapName)
+
+	switch {
+	case !map4Found && !map6Found:
+		return 0, nil
+	case map4Found && maglev4TableSize == UnknownMaglevTableSize:
+		return 0, errors.New("cannot determine v4 outer maglev map's table size")
+	case map6Found && maglev6TableSize == UnknownMaglevTableSize:
+		return 0, errors.New("cannot determine v6 outer maglev map's table size")
+	case map4Found && map6Found && maglev4TableSize != maglev6TableSize:
+		// Just being extra defensive here. This case should never
+		// happen as both maps are created at the same time after
+		// deleting eventual old maps with a different M parameter
+		return 0, errors.New("v4 and v6 maps have different table sizes")
+	case map4Found:
+		detectedTableSize = maglev4TableSize
+	case map6Found:
+		detectedTableSize = maglev6TableSize
+	}
+
+	if map4Found {
+		maglevOuter4Map, err = OpenMaglevOuterMap(MaglevOuter4MapName, detectedTableSize)
+		if err != nil {
+			return UnknownMaglevTableSize, err
 		}
 	}
 
-	if deleteMap {
-		log.WithField(logfields.BPFMapName, mapName).
-			Info("Deleting Maglev outer map due to different M or empty map")
-		if err := m.Unpin(); err != nil {
-			return false, err
+	if map6Found {
+		maglevOuter6Map, err = OpenMaglevOuterMap(MaglevOuter6MapName, detectedTableSize)
+		if err != nil {
+			return UnknownMaglevTableSize, err
 		}
 	}
 
-	return deleteMap, nil
+	return detectedTableSize, nil
 }
 
-func newInnerMaglevMap(name string) *bpf.Map {
-	return bpf.NewMapWithOpts(
-		name,
-		bpf.MapTypeArray,
-		&MaglevInnerKey{}, int(unsafe.Sizeof(MaglevInnerKey{})),
-		&MaglevInnerVal{}, int(unsafe.Sizeof(uint16(0)))*option.Config.MaglevTableSize,
-		1, 0, 0,
-		bpf.ConvertKeyValue,
-		&bpf.NewMapOpts{},
-	)
+// GetOpenMaglevMaps returns a map with all the opened outer maglev eBPF maps.
+// These BPF maps are indexed by their name.
+func GetOpenMaglevMaps() map[string]*maglevOuterMap {
+	maps := map[string]*maglevOuterMap{}
+	if maglevOuter4Map != nil {
+		maps[MaglevOuter4MapName] = maglevOuter4Map
+	}
+	if maglevOuter6Map != nil {
+		maps[MaglevOuter6MapName] = maglevOuter6Map
+	}
+
+	return maps
```

Comment:
Nit: This function seems unnecessary. We can store the maglev outer maps in fixed array and have it as an exported variable. And the maps names are already stored inside the `Map` object anyway, so we can retrieve those at the caller (for-loops) also

Reply:
> And the maps names are already stored inside the Map object anyway

Only when we init/create the maps with `InitMaglevMaps()`. When the maps are loaded from the pinned object (with `OpenMaglevMaps()`, the spec is not populated and so we don't have access to the name

Reply:
Ah OK thanks for clarifying.